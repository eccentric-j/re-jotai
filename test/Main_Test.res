open Jest
open Expect
open ReactTestingLibrary

let initialCounter = 42

let counterAtom = Jotai.atom(initialCounter)

let messageAtom = Jotai.atom("Welcome Jotai!")

let doubleCounterDerivedAtom = Jotai.derivedAtom(get => get(counterAtom) * 2)

let asyncDerivedAtom = Jotai.derivedAsyncAtom(get =>
  Js.Promise.make((~resolve, ~reject as _) => {
    let tripleCounter = get(counterAtom) * 3

    Js.Global.setTimeout(() => resolve(. tripleCounter), 300)->ignore
  })
)

module AsyncCounter = {
  @react.component
  let make = () => {
    // let _tripleCounter = Jotai.useAtom(asyncDerivedAtom)
    let tripleCounter = Jotai.useDerivedAtom(asyncDerivedAtom)

    <> <div title="tripled-counter"> {tripleCounter->React.int} </div> </>
  }
}

module Counter = {
  @react.component
  let make = () => {
    let (counter, setCounter) = Jotai.useAtom(counterAtom)
    let _counter = Jotai.useDerivedAtom(counterAtom)
    // let _doubleCounter = Jotai.useAtom(doubleCounterDerivedAtom)
    let doubleCounter = Jotai.useDerivedAtom(doubleCounterDerivedAtom)

    <div>
      <div title="counter"> {counter->React.int} </div>
      <div title="doubled-counter"> {doubleCounter->React.int} </div>
      <React.Suspense fallback={<div> {"loading"->React.string} </div>}>
        <AsyncCounter />
      </React.Suspense>
      <button title="increment" onClick={_ => setCounter(counter => counter + 1)}>
        {"increment"->React.string}
      </button>
    </div>
  }
}

module App = {
  @react.component
  let make = () => <Jotai.Provider> <Counter /> </Jotai.Provider>
}

describe("useCounter", () => {
  test(`Counter is ${initialCounter->Int.toString}`, () => {
    let app = <App />->render

    app->container->expect->toMatchSnapshot
  })

  test(`Counter is ${(initialCounter + 1)->Int.toString}`, () => {
    let app = <App />->render

    app->getByText(~matcher=#Str("increment"))->FireEvent.click->ignore

    app->container->expect->toMatchSnapshot
  })
})
