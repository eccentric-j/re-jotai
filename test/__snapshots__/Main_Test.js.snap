// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCounter Counter is 42 1`] = `
<div>
  <div>
    <div
      title="counter"
    >
      42
    </div>
    <div
      title="doubled-counter"
    >
      84
    </div>
    <div>
      loading
    </div>
    <button
      title="increment"
    >
      increment
    </button>
  </div>
</div>
`;

exports[`useCounter Counter is 43 1`] = `
<div>
  <div>
    <div
      title="counter"
    >
      43
    </div>
    <div
      title="doubled-counter"
    >
      86
    </div>
    <div>
      loading
    </div>
    <button
      title="increment"
    >
      increment
    </button>
  </div>
</div>
`;
